File: packages/shared/src/dtos/channel.rs
use crate::data::{CategoryId, channel::prelude::ChannelMode};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CreateChannelKind {
    Text,
    Voice,
    Canvas,
    Docs,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateChannelRequest {
    pub name: String,
    pub kind: CreateChannelKind,
    pub category_id: Option<CategoryId>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateChannelRequest {
    pub name: Option<String>,
    pub position: Option<i32>,
    pub category_id: Option<Option<CategoryId>>,
    pub mode: Option<ChannelMode>,
    pub bitrate: Option<i32>,
    pub user_limit: Option<Option<i32>>,
}

File: apps/client/src/types/channel.ts
import type { CategoryId, ChannelId, GuildId, MessageId, UserId } from ".";
import type { UserPublic } from "./user";

export interface CanvasChannel {
  category_id: CategoryId | null;
  guild_id: GuildId;
  id: ChannelId;
  name: string;
  position: number;
}

export interface ChannelCategory {
  guild_id: GuildId;
  id: CategoryId;
  name: string;
  position: number;
}

export interface DMChannel {
  id: ChannelId;
  is_open: boolean;
  recipients: UserPublic[];
}

export interface DocsChannel {
  category_id: CategoryId | null;
  guild_id: GuildId;
  id: ChannelId;
  name: string;
  position: number;
}

export interface Message {
  author_id: UserId;
  channel_id: ChannelId;
  content: string;
  created_at: string;
  deleted: boolean;
  edited_at: string | null;
  id: MessageId;
  thread_id: MessageId | null;
}

export interface PinnedMessage {
  label: string | null;
  message_id: MessageId;
  pinned_at: string;
  pinned_by: UserId;
}

export type ChannelMode = "chat" | "board" | "threads";

export interface TextChannel {
  category_id: CategoryId | null;
  guild_id: GuildId;
  id: ChannelId;
  mode: ChannelMode;
  name: string;
  position: number;
}

export interface VoiceParticipant {
  deafened: boolean;
  muted: boolean;
  speaking: boolean;
  user_id: UserId;
}

export interface VoiceChannel {
  bitrate: number;
  category_id: CategoryId | null;
  guild_id: GuildId;
  id: ChannelId;
  name: string;
  participants: VoiceParticipant[];
  position: number;
  user_limit: number | null;
}

export type Channel =
  | ({ kind: "text" } & TextChannel)
  | ({ kind: "voice" } & VoiceChannel)
  | ({ kind: "docs" } & DocsChannel)
  | ({ kind: "canvas" } & CanvasChannel);
